src/execution.rs

```diff
@@ -114,6 +122,116 @@ pub trait Report {
     fn start(&mut self, task_name: String);
     fn report(&mut self, result: &CompletedTask);
     fn done(&mut self);
+
+    fn stage(&mut self, full_name: &[String], stage_rep: StageReport) {
+        let mut full_name: Vec<String> = Vec::from(full_name);
+        full_name.push(stage_rep.stage_name);
+        let completed_task = CompletedTask {
+            full_name,
+            duration: stage_rep.duration,
+            stdout: Vec::new(),
+            stderr: Vec::new(),
+            status: Status::from(stage_rep.status),
+        };
+        self.report(&completed_task);
+    }
+}
+
+pub struct StageReportSender {
```

Comment:
Let's use a more generic name, something like "TestContext".

There are plenty of use-cases for the control channel beside stage reporting (I actually wanted to have the control channel before, but thought it's an unnecessary complication…), e.g.

EXPECT assertions that don't fail the test immediately:
> The assertions come in pairs that test the same thing but have different effects on the current function. ASSERT_* versions generate fatal failures when they fail, and abort the current function. EXPECT_* versions generate nonfatal failures, which don't abort the current function. Usually EXPECT_* are preferred, as they allow more than one failures to be reported in a test. However, you should use ASSERT_* if it doesn't make sense to continue when the assertion in question fails.
-- https://chenchang.gitbooks.io/googletest_docs/content/googletest/Primer.html

Another use could be a more precise reporting of `should_panic` tests.